[JSC] Implement String.prototype.replace / String.prototype.replaceAll in C++

Constellation opened the pull request:
#### 2b37638c2082b7c545b0dc8a0474d49398538062
<pre>
[JSC] Implement String.prototype.replace / String.prototype.replaceAll in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=304713">https://bugs.webkit.org/show_bug.cgi?id=304713</a>
<a href="https://rdar.apple.com/167219298">rdar://167219298</a>

Reviewed by Sosuke Suzuki.

This patch moves String.prototype.replace and
String.prototype.replaceAll implementations from JS to C++. We use
watchpoint to monitor RegExp getter changes.

                                                     ToT                     Patched

string-replace-generic                         14.8738+-0.0330     ^     14.7730+-0.0546        ^ definitely 1.0068x faster
string-replace                                  1.2715+-0.0153     ^      1.0564+-0.0147        ^ definitely 1.2037x faster
string-replace-one-char-long-string-string-string
                                                0.2625+-0.0044     ^      0.2526+-0.0042        ^ definitely 1.0394x faster
string-replace-one-char-short-string-string-string
                                                0.5615+-0.0046     ^      0.5423+-0.0047        ^ definitely 1.0355x faster
string-replace-benchmark                       54.2826+-0.2489           54.0185+-0.3693
string-replace-string                         227.2664+-3.6705     ^    218.6221+-2.8812        ^ definitely 1.0395x faster
string-replace-empty                            5.1744+-0.0219     ^      4.8294+-0.0189        ^ definitely 1.0714x faster

Disabling DFG and FTL.
                                                     ToT                     Patched

string-replace-generic                         30.7351+-0.0993     ^     16.3372+-0.1908        ^ definitely 1.8813x faster
string-replace                                 31.0990+-0.2213     ^     15.9298+-0.1032        ^ definitely 1.9523x faster
string-replace-one-char-long-string-string-string
                                                0.2481+-0.0041            0.2458+-0.0038
string-replace-one-char-short-string-string-string
                                                0.6327+-0.0042     ^      0.5588+-0.0050        ^ definitely 1.1323x faster
string-replace-benchmark                       72.3342+-0.4864     ^     66.6822+-0.4705        ^ definitely 1.0848x faster
string-replace-string                         349.1991+-3.1613     ^    289.9050+-4.1124        ^ definitely 1.2045x faster
string-replace-empty                           23.2936+-0.0631     ^      7.7031+-0.0961        ^ definitely 3.0239x faster

* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/StringPrototype.js:
(linkTimeConstant.hasObservableSideEffectsForStringReplace): Deleted.
(intrinsic.StringPrototypeReplaceIntrinsic.replace): Deleted.
(intrinsic.StringPrototypeReplaceAllIntrinsic.replaceAll): Deleted.
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/RegExpObject.h:
* Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
(JSC::RegExpObject::isSymbolReplaceFastAndNonObservable):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/304954@main">https://commits.webkit.org/304954@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d9ddfc9a61b4b254acde4a3c54050940b634e71

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9364 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144747 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89977 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6fa13ca0-94b3-48bd-854d-020938efe8e8) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10067 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9491 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104764 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e48145f5-c6cb-462c-810d-74473dc57933) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139949 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7389 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122766 "Passed tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d8bd39da-a699-470b-99a9-5e334d004030) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7026 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4725 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5336 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128965 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116367 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40938 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147502 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135490 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9047 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41507 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9065 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7601 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6948 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119037 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63330 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21114 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9095 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37091 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/168271 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8818 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72661 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43908 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9036 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8888 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
```diff
@@ -30,9 +30,47 @@
+
+    // Otherwise, check that the RegExpObject doesn't have its own Symbol.replace property using
+    // getDirectOffset which doesn't throw.
+    if (getDirectOffset(vm, vm.propertyNames->replaceSymbol) != invalidOffset)
```
Is it sufficient to just check `@@replace`? Don't we need to check the flag getters here as well?

sosukesuzuki replied:
Ah sorry you updated to use `hasCustomProperty` already

sosukesuzuki commented on Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
```diff
@@ -30,9 +30,38 @@
+    if (structure->storedPrototype() != globalObject->regExpPrototype())
+        return false;
+
+    if (!hasCustomProperties())
```
?

```suggestion
    if (hasCustomProperties())
```

Constellation replied:
Right. Changing now.

sosukesuzuki reviewed (approved):
r=me, if EWS is happy

Constellation committed:
[JSC] Implement String.prototype.replace / String.prototype.replaceAll in C++
https://bugs.webkit.org/show_bug.cgi?id=304713
rdar://167219298

Reviewed by Sosuke Suzuki.

This patch moves String.prototype.replace and
String.prototype.replaceAll implementations from JS to C++. We use
watchpoint to monitor RegExp getter changes.

                                                     ToT                     Patched

string-replace-generic                         14.8738+-0.0330     ^     14.7730+-0.0546        ^ definitely 1.0068x faster
string-replace                                  1.2715+-0.0153     ^      1.0564+-0.0147        ^ definitely 1.2037x faster
string-replace-one-char-long-string-string-string
                                                0.2625+-0.0044     ^      0.2526+-0.0042        ^ definitely 1.0394x faster
string-replace-one-char-short-string-string-string
                                                0.5615+-0.0046     ^      0.5423+-0.0047        ^ definitely 1.0355x faster
string-replace-benchmark                       54.2826+-0.2489           54.0185+-0.3693
string-replace-string                         227.2664+-3.6705     ^    218.6221+-2.8812        ^ definitely 1.0395x faster
string-replace-empty                            5.1744+-0.0219     ^      4.8294+-0.0189        ^ definitely 1.0714x faster

Disabling DFG and FTL.
                                                     ToT                     Patched

string-replace-generic                         30.7351+-0.0993     ^     16.3372+-0.1908        ^ definitely 1.8813x faster
string-replace                                 31.0990+-0.2213     ^     15.9298+-0.1032        ^ definitely 1.9523x faster
string-replace-one-char-long-string-string-string
                                                0.2481+-0.0041            0.2458+-0.0038
string-replace-one-char-short-string-string-string
                                                0.6327+-0.0042     ^      0.5588+-0.0050        ^ definitely 1.1323x faster
string-replace-benchmark                       72.3342+-0.4864     ^     66.6822+-0.4705        ^ definitely 1.0848x faster
string-replace-string                         349.1991+-3.1613     ^    289.9050+-4.1124        ^ definitely 1.2045x faster
string-replace-empty                           23.2936+-0.0631     ^      7.7031+-0.0961        ^ definitely 3.0239x faster

* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/StringPrototype.js:
(linkTimeConstant.hasObservableSideEffectsForStringReplace): Deleted.
(intrinsic.StringPrototypeReplaceIntrinsic.replace): Deleted.
(intrinsic.StringPrototypeReplaceAllIntrinsic.replaceAll): Deleted.
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/RegExpObject.h:
* Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
(JSC::RegExpObject::isSymbolReplaceFastAndNonObservable):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/304954@main
